### PR TITLE
Fix libtapi for PowerPC and bump MoarVM/nqp/rakudo versions

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

A fix for Darwin PowerPC.

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.6
Xcode 3.2

###### Verification 

- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL (https://trac.macports.org/ticket/64187)

### CODE REVIEW

**Comment by Gcenx** [comment]
You really need to split these into separate PRs

The commit messages should be something like:
- libtapi: fix powerpc
- MoarVM: updated to 2022.02
- vala: updated to 0.56.0

**Comment by barracuda156** [comment]
> You really need to split these into separate PRs
> 
> The commit messages should be something like:
> 
> * libtapi: fix powerpc
> * MoarVM: updated to 2022.02
> * vala: updated to 0.56.0

@Gcenx I honestly tried to, but Git still pulls these together. If you or anyone can edit these, would be great.

**Comment by mojca** [comment]
Update to MoarVM probably needs an update to rakudo and nqp as well.

**Comment by barracuda156** [comment]
> Update to MoarVM probably needs an update to rakudo and nqp as well.

@mojca It is there, I updated all three.

**Comment by reneeotten** [comment]
> @Gcenx I honestly tried to, but Git still pulls these together. If you or anyone can edit these, would be great.

@barracuda156 the link I gave you in one of the Trac tickets explains quite well how to do these things (i.e., you want to create different branches). Additionally, we do not want to have "merge commits" you will need to do a rebase with master.

Please spend a little bit of time learning the basics of git now (Google is your friend, there are numerous tutorials on this), this will really help you (and us) in the long run.

**Comment by barracuda156** [comment]
> Please spend a little bit of time learning the basics of git now (Google is your friend, there are numerous tutorials on this), this will really help you (and us) in the long run.

@reneeotten This was my first attempt, hope you excuse me. Once I saw it does not work correctly, I will make sure not to repeat the same. (On a side note: it is perhaps one of the most counter-intuitive things UI-wise.)

**Comment by mojca** [comment]
In any case you should really avoid making any changes to the master branch of the repository: that's almost a guarantee that things will sooner or later go wrong unless you know what you are doing.

**Comment by mojca** [comment]
I rebased and reworded the commits to match our style.
(I didn't touch any of the changes as they all look clean/good to me.)

In theory this now looks good to be merged.
Yes, in principle these are three separate contributions that should have ended up in separate pull requests, but let's not overcomplicate now and let's merge it as it is (unless anyone has further feedback that needs to be taken into account).

@barracuda156: please try to avoid pushing random new commits or merges to the master branch of your fork until this gets merge. Once this PR gets merged, it makes sense to create new branches for every single contribution. We can provide further help with git in case you need it.

In any case, thank you very much for your contribution.

**Comment by barracuda156** [comment]
> I rebased and reworded the commits to match our style. (I didn't touch any of the changes as they all look clean/good to me.)
> 
> In theory this now looks good to be merged. Yes, in principle these are three separate contributions that should have ended up in separate pull requests, but let's not overcomplicate now and let's merge it as it is (unless anyone has further feedback that needs to be taken into account).
> 
> @barracuda156: please try to avoid pushing random new commits or merges to the master branch of your fork until this gets merge. Once this PR gets merged, it makes sense to create new branches for every single contribution. We can provide further help with git in case you need it.
> 
> In any case, thank you very much for your contribution.

@mojca Thank you very much! Rest assured, next time I will arrange commits appropriately (and yes, I will make separate branches).

**Comment by evanmiller** [comment]
@mojca I have a few changes in the works for MoarVM (including #14360 and #14336) so I'd prefer to hold off on a version update until those are finished.

**Comment by reneeotten** [comment]
> > Please spend a little bit of time learning the basics of git now (Google is your friend, there are numerous tutorials on this), this will really help you (and us) in the long run.
> 
> @reneeotten This was my first attempt, hope you excuse me. 

certainly - not a problem, it takes a bit of time to get the hang of it!

> Once I saw it does not work correctly, I will make sure not to repeat the same. (On a side note: it is perhaps one of the most counter-intuitive things UI-wise.)

I'd really recommend doing all of this on the command-line and certainly staying away from the GitHub website as it appears one cannot easily accomplish these tasks of rebasing/ammending and such from the GUI.

**Comment by barracuda156** [comment]
> @mojca I have a few changes in the works for MoarVM (including #14360 and #14336) so I'd prefer to hold off on a version update until those are finished.

@evanmiller We may need `--no-mimalloc` flag for PPC (I have it in my local portfile, but it is not there in the Macports repo).

Also, conditional now `if {${os.platform} eq "darwin" && ${os.arch} eq "powerpc"}` is wrong – on Rosetta it fails (flag does not get passed apparently because `os.arch` is not determined as ppc – I guess it should rather be `build.arch` instead):
```
In file included from 3rdparty/mimalloc/src/static.c:23:
3rdparty/mimalloc/src/stats.c: In function 'mi_stat_process_info':
3rdparty/mimalloc/src/stats.c:534:31: error: storage size of 'info' isn't known
  534 |   struct mach_task_basic_info info;
      |                               ^~~~
3rdparty/mimalloc/src/stats.c:535:38: error: 'MACH_TASK_BASIC_INFO_COUNT' undeclared (first use in this function); did you mean 'TASK_BASIC_INFO_COUNT'?
  535 |   mach_msg_type_number_t infoCount = MACH_TASK_BASIC_INFO_COUNT;
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                      TASK_BASIC_INFO_COUNT
3rdparty/mimalloc/src/stats.c:535:38: note: each undeclared identifier is reported only once for each function it appears in
3rdparty/mimalloc/src/stats.c:536:35: error: 'MACH_TASK_BASIC_INFO' undeclared (first use in this function); did you mean 'TASK_BASIC_INFO'?
  536 |   if (task_info(mach_task_self(), MACH_TASK_BASIC_INFO, (task_info_t)&info, &infoCount) == KERN_SUCCESS) {
      |                                   ^~~~~~~~~~~~~~~~~~~~
      |                                   TASK_BASIC_INFO
In file included from 3rdparty/mimalloc/src/static.c:29:
3rdparty/mimalloc/src/segment.c: In function 'mi_tagged_segment_ptr':
3rdparty/mimalloc/src/segment.c:1081:10: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
 1081 |   return (mi_segment_t*)(ts & ~MI_TAGGED_MASK);
      |          ^
In file included from src/moar.h:90,
                 from src/moar.c:2:
src/core/nativecall.h: In function 'MVM_nativecall_find_thread_context':
src/platform/threads.h:19:34: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
   19 | #define MVM_platform_thread_id() (MVMint64)uv_thread_self()
      |                                  ^
src/core/nativecall.h:146:19: note: in expansion of macro 'MVM_platform_thread_id'
  146 |                   MVM_platform_thread_id());
      |                   ^~~~~~~~~~~~~~~~~~~~~~
src/moar.c: In function 'MVM_vm_create_instance':
src/platform/threads.h:19:34: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
   19 | #define MVM_platform_thread_id() (MVMint64)uv_thread_self()
      |                                  ^
src/moar.c:235:48: note: in expansion of macro 'MVM_platform_thread_id'
  235 |     instance->threads->body.native_thread_id = MVM_platform_thread_id();
      |                                                ^~~~~~~~~~~~~~~~~~~~~~
3rdparty/mimalloc/src/static.c: At top level:
cc1: note: unrecognized command-line option '-Wno-logical-op-parentheses' may have been intended to silence earlier diagnostics
make: *** [3rdparty/mimalloc/src/static.o] Error 1
make: *** Waiting for unfinished jobs....
```

P. S. I also don’t get why portfile allows `macport-gcc-6` and _down_ for compilers. `gcc7` exists for PPC for ages, and now both `gcc10` and `gcc11` build and work.

**Comment by evanmiller** [comment]
@barracuda156 All the `mimalloc` stuff is new in 2022.02; it's not clear to me whether it won't run on PPC, or is instead restricted to certain versions of macOS. It could require 10.7+; I'm not sure. I believe the `if {${os.platform} eq "darwin" && ${os.arch} eq "powerpc"}` clause can be ripped out, but this will need testing.

I will repeat my preference to sort out the legacy systems using the current version of MoarVM, and then look into upgrading. That will give us a working baseline without the `mimalloc` complications. To that end I am waiting on #14366 to be reviewed and merged.

**Comment by evanmiller** [comment]
Okay my last Raku-related PR is #14449. That gets Tiger/PPC working, then you have my blessing to upgrade.

**Comment by barracuda156** [comment]
> Okay my last Raku-related PR is #14449. That gets Tiger/PPC working, then you have my blessing to upgrade.

@evanmiller Thank you! (I cannot push the commit into master repo, so let’s wait for someone.)

**Review by mascguy — APPROVED** [review]

**Comment by mascguy** [comment]
@barracuda156 I went ahead and fixed the merge conflict for `moarvm`, along with adding revisions for `nqp` and `rakudo`.

Once the CI builds complete successfully, I'll merge.

**Comment by evanmiller** [comment]
`--no-mimalloc` will need to be passed to MoarVM's configure on 10.6 and earlier.

I really think it's worth splitting this into separate PRs.

**Comment by mascguy** [comment]
It looks like the CI builds failed as well, though I haven't had a chance to review yet.

**Comment by mascguy** [comment]
Rebased against master

**Comment by barracuda156** [comment]
@mascguy `libtapi` commit not yet pushed, right? https://github.com/macports/macports-ports/pull/14359/commits/41c5870e2c9d6058e0bb334ee721114f57b660ac

**Comment by mascguy** [comment]
> @mascguy `libtapi` commit not yet pushed, right? [41c5870](https://github.com/macports/macports-ports/commit/41c5870e2c9d6058e0bb334ee721114f57b660ac)

Correct, that change should still be part of this PR.

**Comment by mojca** [comment]
Commits need a cleanup.

**Comment by barracuda156** [comment]
> Commits need a cleanup.

@mojca Sure, what should I fix about https://github.com/macports/macports-ports/commit/41c5870e2c9d6058e0bb334ee721114f57b660ac ?

**Comment by mascguy** [comment]
> > Commits need a cleanup.
> 
> @mojca Sure, what should I fix about [41c5870](https://github.com/macports/macports-ports/commit/41c5870e2c9d6058e0bb334ee721114f57b660ac) ?

You added a merge commit with your last push. But that wasn't needed, as I had just rebased against master - and pushed - a short while prior to yours. (In short, your push was redundant.) No worries, but just an FYI.

I just force-pushed again, without a merge commit. Don't force-push again!  ;-)

**Comment by mascguy** [comment]
> As I mentioned earlier, `--no-mimalloc` needs to be passed to MoarVM on 10.6 and earlier.
> 
> There is already a newer version of MoarVM / nqp / rakudo out (2022.03). I will reiterate my suggestion to break this PR into separate ones to facilitate discussion and merging.

I just took care of updating `vala` on master, so that's no longer needed in this PR. Removed from the commits and force-pushed.

So that simplifies things a bit more. What else do we need to split out?

**Comment by mojca** [comment]
It looks ok to me in any case.

**Comment by mascguy** [comment]
Just squashed the revision-related additions, so that they're folded into the other updates. (Makes things easier for any other minor additions.)

**Comment by mascguy** [comment]
> As I mentioned earlier, `--no-mimalloc` needs to be passed to MoarVM on 10.6 and earlier.

Evan, I can quickly add this change. Is this simply an additional configure argument, for darwin >= 10?

**Comment by mascguy** [comment]
> > As I mentioned earlier, `--no-mimalloc` needs to be passed to MoarVM on 10.6 and earlier.
> 
> Evan, I can quickly make this change. Is this simply an additional configure argument, for darwin >= 10?

Added via additional commit. Let me know if this covers it.

**Comment by barracuda156** [comment]
@mascguy Thank you! Sorry again for this mess, this was my first attempt, and it wasn't smooth. I will keep things separate in the future.
Now I just need to get to the stage that my fork has no commits ahead of Macports master.

**Comment by mascguy** [comment]
> @mascguy Thank you! Sorry again for this mess, this was my first attempt, and it wasn't smooth. I will keep things separate in the future. Now I just need to get to the stage that my fork has no commits ahead of Macports master.

No worries at all! We're just trying to assist, and help you move this along.

And we really appreciate your initiative, relative all of your contributions to MacPorts... not just this PR, but everything. Much appreciated!

**Comment by mascguy** [comment]
Tested locally, and looks good. So let's merge this round of changes, and release the hounds!

**Comment by mascguy** [comment]
And great job @barracuda156 !

**Comment by barracuda156** [comment]
> And great job @barracuda156 !

@mascguy Thank you!

P. S. I have few port fixes for PPC in “works-for-me” stage, will review to make sure they are neat and then commit.

File: lang/MoarVM/Portfile
Comment (evanmiller) [inline lang/MoarVM/Portfile:57]: Should be `<= 10` but we can fix this in the next version bump

Comment (mascguy) [reply lang/MoarVM/Portfile:57]: Thanks for the heads-up; fixed, and rev-bumped.

**Pull request merged**
